fix(rust,python): Fix precision/scale handling and invalid numbers in string-to-decimal conversions.

cgevans opened the pull request:
Closes #13546. Closes #13568.  Also fixes bug found in process of writing this, where, eg, "12.-2" is parsed as 11.98.  Checks precision and scale in what I believe is the correct way, and simplifies the code somewhat.  Removes comments that appeared to be either wrong or confusing.

Original description is below.

---

fix: Parse string without "." as decimal even with scale > 0.

Note, however, that I'm confused by some behavior of `deserialize_decimal`, and it seems like it might be worth changing it further while looking at it.  Given the way decimals are stored (in an i128), my assumption was that the scale would be, essentially, the multiplier ($10^{scale}$), while the precision would be the maximum total number of digits.  A comment states that the function doesn't check precision or scale.  Yet it appears:

- Precision is checked against (lhs digits - leading lhs zeros) + (rhs digits - leading rhs zeros), when it seems like it should be checked against (lhs digits - leading lhs zeros) + (rhs digits), or  (lhs digits - leading lhs zeros) + max(rhs digits, scale).  The current arrangement means that `5000.002` is considered to fit in precision 5, scale 4, for example.
- Scale is checked against (rhs digits - leading rhs zeros), when it seems like it should be checked against rhs digits.

cgevans commented:
Actually, this brings up an additional bug:

```
>>> import polars as pl
>>> pl.Series(["12.-2"]).cast(pl.Decimal(scale=5))
shape: (1,)
Series: '' [decimal[*,5]]
[
	11.98000
]
```

ritchie46 commented:
I always have to look up the spec again when dealing with decimal.  Thanks for the fix. Can you leave some comments explaining your rationale?

cgevans commented:
I've added some explanatory comments.  Adding these led down a path to finding another bug caused by `significant_figures` being used where it shouldn't have been (#13568), so I have additionally fixed that.  In doing so, I've just removed `significant_figures` entirely, as there was only one place where it was actually the right calculation to make.  Adding a test for #13568 required adding `PartialEq` for `AnyValue::Decimal`.  I haven't added any other handling in `AnyValue`.

I'm not entirely about the legacy vs non-legacy distinction in polars-arrow, so I'm not sure whether this code should be moved.

cgevans commented:
I don't understand the test failure here (in `tests::io::test_ndjson_globbing` in `polars-lazy`).  It appears to be entirely unrelated to the PR, and does not happen on my laptop (Fedora 39) or the windows test action.

alexander-beedie commented:
> I don't understand the test failure here 

There has been a lot of GitHub instability this afternoon; I'll rerun the failed job and see if that clears it 🤔

cgevans commented:
Apparently rebasing fixed it.

ritchie46 commented:
Thank you for the additional comments! Our decimial support is not near where we want it to be. So this is super helpful. Thanks a lot.
